src/stan/services/pathfinder/single.hpp

```diff
@@ -857,111 +845,133 @@ inline auto pathfinder_lbfgs_single(
+          constrained_draws_vec.head(2) = elbo_lp_mat.row(i);
+          unconstrained_col = elbo_draws.col(i);
+          constrained_draws_vec.tail(num_unconstrained_params)
+              = constrain_fun(rng, unconstrained_col,
+                              approx_samples_constrained_col)
+                    .matrix();
+          parameter_writer(constrained_draws_vec);
+        }
+        lp_ratio = std::move(elbo_best.lp_ratio);
       }
-    } catch (const std::domain_error& e) {
-      std::string err_msg = e.what();
-      logger.warn(path_num + "Final sampling approximation failed with error: "
-                  + err_msg);
-      logger.info(
-          path_num
-          + "Returning the approximate samples used for ELBO calculation: "
-          + err_msg);
-      constrained_draws_mat
-          = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>(
-              names.size(), elbo_draws.cols());
+    } else {
+      // output only first num_draws from what we computed for ELBO
       Eigen::VectorXd approx_samples_constrained_col;
       Eigen::VectorXd unconstrained_col;
-      for (Eigen::Index i = 0; i < elbo_draws.cols(); ++i) {
-        constrained_draws_mat.col(i).head(2) = elbo_lp_mat.row(i).matrix();
+      for (Eigen::Index i = 0; i < num_draws; ++i) {
+        constrained_draws_vec.head(2) = elbo_lp_mat.row(i);
         unconstrained_col = elbo_draws.col(i);
-        constrained_draws_mat.col(i).tail(num_unconstrained_params)
+        constrained_draws_vec.tail(num_unconstrained_params)
             = constrain_fun(rng, unconstrained_col,
                             approx_samples_constrained_col)
                   .matrix();
+        parameter_writer(constrained_draws_vec);
       }
-      lp_ratio = std::move(elbo_best.lp_ratio);
+      lp_ratio = std::move(elbo_best.lp_ratio.head(num_draws));
     }
-  } else {
-    // output only first num_draws from what we computed for ELBO
-    constrained_draws_mat
-        = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>(names.size(),
-                                                                num_draws);
-    Eigen::VectorXd approx_samples_constrained_col;
-    Eigen::VectorXd unconstrained_col;
-    for (Eigen::Index i = 0; i < num_draws; ++i) {
-      constrained_draws_mat.col(i).head(2) = elbo_lp_mat.row(i).matrix();
-      unconstrained_col = elbo_draws.col(i);
-      constrained_draws_mat.col(i).tail(num_unconstrained_params)
-          = constrain_fun(rng, unconstrained_col,
-                          approx_samples_constrained_col)
-                .matrix();
+    const auto end_pathfinder_time = std::chrono::steady_clock::now();
+    const double pathfinder_delta_time = stan::services::util::duration_diff(
+        start_pathfinder_time, end_pathfinder_time);
+    // For multi pathfinder, multi would write multiple end times
+    if constexpr (stan::callbacks::is_tee_writer_v<ParamWriter>) {
```

Comment:
I'd prefer depending on a named tparam (even if it's the same as the ReturnLP one, which I think it is?)

Reply:
I changed this to depend on ReturnLP and added a static check for this section that we are working with a tee writer